### ISSUE TITLE
Improve block device CI test

### ISSUE THREAD

**Issue by alexandruag** [rust-vmm/vmm-reference#47]
Run something like `lsblk --json` and check output data such as device name and read-only status. Original proposal in [this](https://github.com/rust-vmm/vmm-reference/pull/41/commits/ac4a9dd45a7b8ce1d299634db3bb96d2a249d483#r540970103) comment.

**Comment by gabhijit** [comment]
I can implement this if no one is taking this. 

I am wondering if it's a good idea to implement by hand a functionality equivalent to `Popen.communicate` while trying to implement fix for this. 

Right now with `expect_string` we can execute only one command (by sending a `bytestring` to the `stdin` like we are doing in `expect_vcpu`), but for executing more than one commands current `expect_string` alone won't work. So something like 

```
def run_cmd_in_vm(cmd_input, timeout):
    # returns the output of running the command or raise Exception.
``` 

Which can allow to run any number of commands once the VM is booted. 

@andreeaflorescu @lauralt what are your thoughts?

**Comment by lauralt** [comment]
Yup, having some sort of abstraction for running a command was something that we were thinking about too. We would've also liked to have something that abstracts away the mechanism by which we send the command to the vm (e.g. stdin, ssh once we merge #49 ). Andreea had more ideas around this.

**Comment by gabhijit** [comment]
I think with `ssh` (once the net functionality is in place), it's relatively easier, one can simply use something like [`paramiko`](https://www.paramiko.org/) (assuming an `sshd` runs inside the VM). 

I will send a PR that implements a basic abstraction for running command while and a fix for this issue.

**Comment by andreeaflorescu** [comment]
We were considering using `pexect` for checking the output of commands. It would be nice to have an abstraction over the channel used for sending the commands because we would like to be able to test the vmm-reference with and without network which means that we cannot always count on the ssh capability to be available.

I am still thinking about the best abstraction, but I was looking for something as follows (we need better names):
-  define a `Channel` trait that has an `execute` command
- Implement the `Channel` trait for `SerialConsole` and for `Ssh`
- the communication `Channel` can be passed to the tests, which then use the `execute` function to check the expected behavior
- in the implementation of the SSH class we can use paramiko

With this approach we can run the same tests with the serial console, or with SSH which is useful for testing multiple flavors of the vmm.

**Comment by gabhijit** [comment]
> 
> With this approach we can run the same tests with the serial console, or with SSH which is useful for testing multiple flavors of the vmm.

After thinking a bit more about it, I am not very sure whether it's required to have the ability to be able to `ssh` to the VM from the host. To test that the networking works, simply being able to `ping` should be good enough. Also being able to `ping` to the host from inside the VM (run through something `pexpect` ish) is good enough. 

It might be required to run some kind of `iperf` test to test the performance of `virtio-net` device. However even that can be achieved by running `iperf` through `stdin` using something `pexpect`ish as above. 

Or am I missing something?

I am considering to send a `PR` using a `run_command` as a fix for this issue, with an abstraction to run any command and check the output. With some code, it might give some idea about what real abstractions would make more sense.

**Comment by andreeaflorescu** [comment]
One reason for running commands with SSH when it's available, it's that SSH is much faster than the serial console. The serial console just sends one byte at a time.

**Comment by andreeaflorescu** [comment]
Fixed by: https://github.com/rust-vmm/vmm-reference/pull/66
